File: macos/Gutenkern/Sources/GutenkernCore/GlyphScript.swift
import Foundation

enum GlyphScript: Equatable, Sendable {
    case latin
    case cyrillic
    case greek
    case unknown

    static func fromUnicodeName(_ unicodeName: String) -> GlyphScript {
        guard let first = unicodeName.split(whereSeparator: { $0.isWhitespace }).first else {
            return .unknown
        }
        switch first {
        case "LATIN", "FEMININE", "MASCULINE":
            return .latin
        case "CYRILLIC":
            return .cyrillic
        case "GREEK":
            return .greek
        default:
            return .unknown
        }
    }

    static func fromCodePoint(_ value: UInt32) -> GlyphScript {
        if (0x0370...0x03FF).contains(value) || (0x1F00...0x1FFF).contains(value) {
            return .greek
        }
        if (0x0400...0x052F).contains(value)
            || (0x2DE0...0x2DFF).contains(value)
            || (0xA640...0xA69F).contains(value)
        {
            return .cyrillic
        }
        if (0x0000...0x02AF).contains(value)
            || (0x1E00...0x1EFF).contains(value)
            || (0x2C60...0x2C7F).contains(value)
            || (0xA720...0xA7FF).contains(value)
            || (0xAB30...0xAB6F).contains(value)
        {
            return .latin
        }
        return .unknown
    }
}

enum KerningScriptFilter {
    static func isLetterGroup(_ group: KerningGroup) -> Bool {
        switch group {
        case .capitals, .smallCaps, .lowercase:
            return true
        case .punctuation, .nonAlphabetic, .liningFigures, .oldstyleFigures:
            return false
        }
    }

    static func shouldKern(_ left: Glyph, _ right: Glyph, letterLetterRecipe: Bool) -> Bool {
        guard letterLetterRecipe else {
            return true
        }
        let leftScript = script(of: left)
        let rightScript = script(of: right)
        if leftScript == .unknown || rightScript == .unknown {
            return true
        }
        return leftScript == rightScript
    }

    static func pairCount(left: [Glyph], right: [Glyph], letterLetterRecipe: Bool) -> Int {
        if !letterLetterRecipe {
            return left.count * right.count
        }
        var total = 0
        for leftGlyph in left {
            for rightGlyph in right where shouldKern(leftGlyph, rightGlyph, letterLetterRecipe: true) {
                total += 1
            }
        }
        return total
    }

    private static func script(of glyph: Glyph) -> GlyphScript {
        switch glyph {
        case .character(let character):
            return GlyphDictionary.script(byChar: String(character))
        case .name(let name):
            return GlyphDictionary.script(byName: name)
        }
    }
}

File: windows/Gutenkern.Core/GlyphScript.cs
namespace Gutenkern;

internal enum GlyphScript
{
    Latin,
    Cyrillic,
    Greek,
    Unknown
}

internal static class KerningScriptFilter
{
    public static bool IsLetterGroup(KerningGroup group) =>
        group is KerningGroup.Capitals or KerningGroup.SmallCaps or KerningGroup.Lowercase;

    public static bool ShouldKern(Glyph left, Glyph right, bool letterLetterRecipe)
    {
        if (!letterLetterRecipe)
        {
            return true;
        }

        var leftScript = ScriptOf(left);
        var rightScript = ScriptOf(right);
        if (leftScript == GlyphScript.Unknown || rightScript == GlyphScript.Unknown)
        {
            return true;
        }

        return leftScript == rightScript;
    }

    public static int PairCount(IReadOnlyList<Glyph> left, IReadOnlyList<Glyph> right, bool letterLetterRecipe)
    {
        if (!letterLetterRecipe)
        {
            return left.Count * right.Count;
        }

        var total = 0;
        foreach (var leftGlyph in left)
        {
            foreach (var rightGlyph in right)
            {
                if (ShouldKern(leftGlyph, rightGlyph, true))
                {
                    total++;
                }
            }
        }

        return total;
    }

    public static GlyphScript FromUnicodeName(string unicodeName)
    {
        if (string.IsNullOrEmpty(unicodeName))
        {
            return GlyphScript.Unknown;
        }

        var first = unicodeName.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        if (first.Length == 0)
        {
            return GlyphScript.Unknown;
        }

        return first[0] switch
        {
            "LATIN" or "FEMININE" or "MASCULINE" => GlyphScript.Latin,
            "CYRILLIC" => GlyphScript.Cyrillic,
            "GREEK" => GlyphScript.Greek,
            _ => GlyphScript.Unknown
        };
    }

    public static GlyphScript FromCodePoint(int value)
    {
        if (value is >= 0x0370 and <= 0x03FF or >= 0x1F00 and <= 0x1FFF)
        {
            return GlyphScript.Greek;
        }

        if (value is >= 0x0400 and <= 0x052F or >= 0x2DE0 and <= 0x2DFF or >= 0xA640 and <= 0xA69F)
        {
            return GlyphScript.Cyrillic;
        }

        if (value is >= 0x0000 and <= 0x02AF
            or >= 0x1E00 and <= 0x1EFF
            or >= 0x2C60 and <= 0x2C7F
            or >= 0xA720 and <= 0xA7FF
            or >= 0xAB30 and <= 0xAB6F)
        {
            return GlyphScript.Latin;
        }

        return GlyphScript.Unknown;
    }

    private static GlyphScript ScriptOf(Glyph glyph) => glyph switch
    {
        CharacterGlyph character => GlyphDictionary.ScriptByChar(character.Value),
        NameGlyph name => GlyphDictionary.ScriptByName(name.Value),
        _ => GlyphScript.Unknown
    };
}
